BUG: Fix csv.QUOTE_NONNUMERIC quoting in to_csv

gfyoung opened the pull request:
Closes #12922: "bug" traced to #12194 via bisection, where the float formatting was unconditionally casting everything to string.

I say "bug" (with quotations) because the changes to `get_result_as_array` did **correctly** cast everything to string as per the documentation (i.e. it had inadvertently patched a bug itself even though it was just a cleaning PR).  However, the changes had overlooked the impact it would have on `to_csv`.

gfyoung committed:
BUG: Fix quoting behaviour in to_csv for csv.QUOTE_NONNUMERIC

Float values were being quoted despite the quoting spec.
Bug traced to the float formatting that was unconditionally
casting all floats to string. Unconditional casting traced
back to commit 2d51b33 (gh-12194) via bisection. This commit
undoes some of those changes to rectify the behaviour.

Closes gh-12922.

[ci skip]

jreback commented on pandas/formats/format.py:
```diff
@@ -2138,6 +2016,31 @@ def formatter(value):
 
+    def _get_formatted_data(self):
+        """
+        Returns formatted float values. This is different from
```
why are you making this a separate function ? the point was to eliminate this kind of thing;

gfyoung replied:
I think my reasoning was made clear in the doc-string.  `get_result_as_array` returns only strings (it supposed to be doing this for quite some time).  However, formatting CSV output is **not always** string.

gfyoung replied:
The point you refer to was well-intended but ultimately incorrect, and sure, we could add this to `get_result_as_array`, but that's an API change then.

jreback replied:
how so? it's an internal routine

gfyoung replied:
`get_result_as_array` - I wouldn't know that from the signature (no preceding `_`)

jreback replied:
these entire module is internal - in reality none should be leading _

we have very inconsistent usage

gfyoung replied:
Ah, I see.  Fair enough, then a note should be made about that in the module doc-string.  I'll move this functionality then into `get_result_as_array` then.

jreback replied:
that's fine

gfyoung replied:
Added note to `format.py` - unfortunately, integrating the no-format option from my new function into `get_result_as_array` caused a whole slew of test failures.  So I just moved it into `to_native_types` instead.

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -101,6 +101,7 @@ API changes
 ~~~~~~~~~~~
 
 
+- ``DataFrame.to_csv()`` has dropped the ``engine`` parameter (:issue:`13418`)
```
this is a prior deprecation removal

do in s separate pr

gfyoung replied:
It made it easier to see what I was doing when I was fixing the bug in `to_csv` though (didn't have to consider the other `engine`).

jreback replied:
I get it but should do that change for 0.19 (should have done it for 0.18 but that ship sailed)

gfyoung replied:
No API changes like this in a minor release?  It has been deprecated for a long time though...do changes like these only go for major releases?

jreback replied:
it's a prior deprecation removal / so no

that would totally break if you had ignored the warning and not fixed your code

people should have done it - but can't hurt to put in major version

gfyoung replied:
True, though the ignoring warning and breaking all of my code argument could be made for a major release as well...but fair enough: will break this up.

gfyoung commented:
@jreback : Moved `engine` parameter removal into separate PR and removed the `_get_formatted_data` function and move its code in `to_native_types`.  Travis is still happy, so ready to merge if there are no other concerns.

jreback commented on pandas/formats/format.py:
```diff
@@ -1,4 +1,8 @@
 # -*- coding: utf-8 -*-
+"""
+Internal module for formatting output data (e.g. `to_csv`).
+"""
```
this is not a very helpful comment.

gfyoung replied:
What should I add exactly?

jreback replied:
you can remove the ``to_csv`` (or add that it provides formatters for csv, html, latex and well as display formatting). It doesn't support ``to_csv``, rather the implemenation for it.

gfyoung replied:
Sure thing.  Done.

jreback commented on pandas/core/internals.py:
```diff
@@ -1529,6 +1529,18 @@ def to_native_types(self, slicer=None, na_rep='', float_format=None,
             values = values[:, slicer]
 
+        # no special formatting is desired
+        if float_format is None and decimal == '.':
```
I guess quoting *only* matters in a csv (and not generally to output, which is where the Formatters matter). pls add a comment with the issue number

gfyoung replied:
Yep, done.

gfyoung commented:
@jreback : Modified the relevant doc-strings and comments as requested.  Ready to merge if there are no other concerns.

gfyoung commented:
@jreback : Modified the relevant doc-strings and comments as requested.  Ready to merge if there are no other concerns.

TomAugspurger commented on pandas/tests/frame/test_to_csv.py:
```diff
@@ -1131,3 +1102,83 @@ def test_to_csv_with_dst_transitions(self):
+            'c_bool': [True, False],
+        })
+
+        expected = """\
```
Style nit, you can use

```
expected = textwrap.dedent(
    """\
    ,c_bool,c_float,c_int,c_string
    0,True,1.0,42.0,a
    1,False,3.2,,"b,c"
    """
)
```

to avoid the odd indentation. Not worth holding up a merge though, if you aren't able to change.

gfyoung replied:
Odd indentation is the MO it seems for anything CSV-related.  I'll leave for now unless there are really strong objections.  But thanks for mentioning though @TomAugspurger !

jreback commented:
ty
